File: csharp/Solitaire/Klondike/Klondike.cs
using Silnith.Game.Deck;
using Silnith.Game.Klondike.Move;
using Silnith.Game.Klondike.Move.Filter;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Silnith.Game.Klondike
{
    /// <summary>
    /// An implementation of Klondike solitaire.
    /// </summary>
    public class Klondike : IGame<ISolitaireMove, Board>
    {
        private readonly IEnumerable<ISolitaireMoveFilter> filters = new List<ISolitaireMoveFilter>()
        {
            new MoveCapFilter(150),
            new KingMoveMustExposeFaceDownCardFilter(),
            new StockPileRecycleMustBeFollowedByAdvance(),
            new StockPileAdvanceMustBeFollowedBySomethingUseful(),
            new DrawFromFoundationMustBeUsefulFilter(),
            new DrawFromStockPileFilter(),
            new RunMoveMustBeFollowedBySomethingUsefulFilter(),
            new BoardCycleFilter(),
        };

        /// <summary>
        /// The number of columns on the board for this game of Klondike solitaire.
        /// </summary>
        /// <remarks>
        /// <para>
        /// This is always <c>7</c>.
        /// </para>
        /// </remarks>
        public int ColumnCount
        {
            get;
        }

        /// <summary>
        /// The number of cards to advance the stock pile.
        /// </summary>
        /// <remarks>
        /// <para>
        /// This is typically <c>3</c>, but could be <c>1</c>.
        /// </para>
        /// </remarks>
        public int DrawAdvance
        {
            get;
        }

        /// <summary>
        /// Constructs a new game instance for exploring the search space of Klondike solitaire.
        /// </summary>
        public Klondike()
        {
            ColumnCount = 7;
            DrawAdvance = 3;
        }

        /// <summary>
        /// Returns all the legal moves for the given board.
        /// </summary>
        /// <param name="board">The board to examine for legal moves.</param>
        /// <returns>All the legal moves.</returns>
        public IEnumerable<ISolitaireMove> FindAllMoves(Board board)
        {
            return Array.Empty<ISolitaireMove>()
                .Concat(StockPileRecycleMove.FindMoves(board))
                .Concat(StockPileAdvanceMove.FindMoves(DrawAdvance, board))
                .Concat(FoundationToColumnMove.FindMoves(board))
                .Concat(ColumnToColumnMove.FindMoves(board))
                .Concat(StockPileToColumnMove.FindMoves(board))
                .Concat(ColumnToFoundationMove.FindMoves(board))
                .Concat(StockPileToFoundationMove.FindMoves(board));
        }

        /// <inheritdoc/>
        public IEnumerable<ISolitaireMove> FindAllMoves(IReadOnlyList<GameState<ISolitaireMove, Board>> state)
        {
            return FindAllMoves(state[0].Board);
        }

        /// <summary>
        /// Returns whether the given board is a winning board for this game.
        /// </summary>
        /// <param name="board">The board to check.</param>
        /// <returns><see langword="true"/> if the board represents a win.</returns>
        public bool IsWin(Board board)
        {
            IReadOnlyDictionary<Suit, IReadOnlyList<Card>> foundation = board.Foundation;
            return foundation[Suit.Club].Count == 13
                && foundation[Suit.Diamond].Count == 13
                && foundation[Suit.Heart].Count == 13
                && foundation[Suit.Spade].Count == 13;
        }

        /// <summary>
        /// Returns whether the given game state is a winning game state for this game.
        /// </summary>
        /// <param name="state">The game state to check.</param>
        /// <returns><see langword="true"/> if the game state represents a win.</returns>
        public bool IsWin(GameState<ISolitaireMove, Board> state)
        {
            return IsWin(state.Board);
        }

        /// <inheritdoc/>
        public bool IsWin(IReadOnlyList<GameState<ISolitaireMove, Board>> gameStates)
        {
            return IsWin(gameStates[0]);
        }

        /// <inheritdoc/>
        public IEnumerable<IMoveFilter<ISolitaireMove, Board>> GetFilters()
        {
            return filters;
        }
    }
}

File: cpp/Solitaire/Klondike/silnith/game/solitaire/klondike.cpp
#include <silnith/game/solitaire/klondike.h>

#include <silnith/game/solitaire/move/ColumnToColumnMove.h>
#include <silnith/game/solitaire/move/ColumnToFoundationMove.h>
#include <silnith/game/solitaire/move/FoundationToColumnMove.h>
#include <silnith/game/solitaire/move/StockPileAdvanceMove.h>
#include <silnith/game/solitaire/move/StockPileRecycleMove.h>
#include <silnith/game/solitaire/move/StockPileToColumnMove.h>
#include <silnith/game/solitaire/move/StockPileToFoundationMove.h>

#include <silnith/game/solitaire/move/filter/BoardCycleFilter.h>
#include <silnith/game/solitaire/move/filter/DrawFromFoundationMustBeUsefulFilter.h>
#include <silnith/game/solitaire/move/filter/DrawFromStockPileFilter.h>
#include <silnith/game/solitaire/move/filter/KingMoveMustExposeFaceDownCardFilter.h>
#include <silnith/game/solitaire/move/filter/MoveCapFilter.h>
#include <silnith/game/solitaire/move/filter/RunMoveMustBeFollowedBySomethingUsefulFilter.h>
#include <silnith/game/solitaire/move/filter/StockPileAdvanceMustBeFollowedBySomethingUsefulFilter.h>
#include <silnith/game/solitaire/move/filter/StockPileRecycleMustBeFollowedByAdvanceFilter.h>

using namespace silnith::game;
using namespace silnith::game::deck;
using namespace silnith::game::solitaire;
using namespace silnith::game::solitaire::move;

using namespace std;

klondike::klondike(void)
    : filters{
        make_shared<move::filter::MoveCapFilter>(150),
        make_shared<move::filter::KingMoveMustExposeFaceDownCardFilter>(),
        make_shared<move::filter::StockPileRecycleMustBeFollowedByAdvanceFilter>(),
        make_shared<move::filter::StockPileAdvanceMustBeFollowedBySomethingUsefulFilter>(),
        make_shared<move::filter::DrawFromFoundationMustBeUsefulFilter>(),
        make_shared<move::filter::DrawFromStockPileFilter>(),
        make_shared<move::filter::RunMoveMustBeFollowedBySomethingUsefulFilter>(),
        make_shared<move::filter::BoardCycleFilter>(),
    }
{}

bool klondike::is_win(board const& board) const
{
    constexpr size_t const thirteen{ static_cast<std::size_t>(value::king) };
    map<suit, vector<card>> const& foundation{ board.get_foundation() };
    if (foundation.at(suit::club).size() < thirteen)
    {
        return false;
    }
    if (foundation.at(suit::diamond).size() < thirteen)
    {
        return false;
    }
    if (foundation.at(suit::heart).size() < thirteen)
    {
        return false;
    }
    if (foundation.at(suit::spade).size() < thirteen)
    {
        return false;
    }
    return true;
}

bool klondike::is_win(shared_ptr<linked_node<game_state<solitaire_move, board>> const> const& node_ptr) const
{
    linked_node<game_state<solitaire_move, board>> const& node{ *node_ptr };
    game_state<solitaire_move, board> const& game_state{ node.get_value() };
    shared_ptr<board const> const& b{ game_state.get_board() };
    return is_win(*b);
}

vector<shared_ptr<solitaire_move const>> klondike::find_all_moves(
    shared_ptr<linked_node<game_state<solitaire_move, board>> const> const& game_state_history) const
{
    game_state<solitaire_move, board> const& game_state{ game_state_history->get_value() };
    shared_ptr<board const> const& board_ptr{ game_state.get_board() };
    board const& b{ *board_ptr };

    vector<shared_ptr<solitaire_move const>> moves{};
    for (shared_ptr<solitaire_move const> move : StockPileRecycleMove::find_moves(b))
    {
        moves.emplace_back(move);
    }
    for (shared_ptr<solitaire_move const> move : StockPileAdvanceMove::find_moves(draw_advance, b))
    {
        moves.emplace_back(move);
    }
    for (shared_ptr<solitaire_move const> move : FoundationToColumnMove::find_moves(b))
    {
        moves.emplace_back(move);
    }
    for (shared_ptr<solitaire_move const> move : ColumnToColumnMove::find_moves(b))
    {
        moves.emplace_back(move);
    }
    for (shared_ptr<solitaire_move const> move : StockPileToColumnMove::find_moves(b))
    {
        moves.emplace_back(move);
    }
    for (shared_ptr<solitaire_move const> move : ColumnToFoundationMove::find_moves(b))
    {
        moves.emplace_back(move);
    }
    for (shared_ptr<solitaire_move const> move : StockPileToFoundationMove::find_moves(b))
    {
        moves.emplace_back(move);
    }
    return moves;
}

span<shared_ptr<move_filter<solitaire_move, board> const> const> klondike::get_filters(void) const
{
    return filters;
}
